drivers/SmartThings/deepsmart/src/client/httpclient.lua

```diff
@@ -0,0 +1,251 @@
+local log = require('log')
+local cosock = require "cosock"
+local http = cosock.asyncify "socket.http"
+local ltn12 = require('ltn12')
+local RestClient = require "lunchbox.rest"
+local utils = require "utils.utils"
+local json = require('dkjson')
+
+local HttpClient = {}
+HttpClient.__index = HttpClient
+---------------
+--------------------
+local SSL_CONFIG = {
+  mode = "client",
+  protocol = "any",
+  verify = "peer",
+  options = "all",
+  cafile = "./selfSignedRoot.crt"
+}
+
+local ADDITIONAL_HEADERS = {
+  ["Accept"] = "application/json",
+  ["Content-Type"] = "application/json",
+}
+
+function HttpClient.labeled_socket_builder(label)
+  local socket_builder = utils.labeled_socket_builder(label, SSL_CONFIG)
+  return socket_builder
+end
+
+
+function HttpClient.client(wiser_index_code, ip)
+  local ret = setmetatable({
+    wiser_index_code = wiser_index_code,
+    ip = ip,
+    client = RestClient.new("https://"..ip, HttpClient.labeled_socket_builder(wiser_index_code))
+  }, HttpClient)
+
+  return ret
+end
+
+local function process_rest_response(response, err, partial)
+  if err ~= nil then
+    return response, err, nil
+  elseif response ~= nil then
+    return response:get_body(), nil, response.status
+  else
+    return nil, "no response or error received", nil
+  end
+end
+
+local function retry_fn(retry_attempts)
+  local count = 0
+  return function()
+    count = count + 1
+    return count < retry_attempts
+  end
+end
+
+function HttpClient:do_get(url)
+  -- get url
+  local client = self.client
+  if (client == nil) then
+    log.warn('do_get url '..url..' client is nil')
+  end
+  log.debug('do_get '..url)
+  local response,err,partial = client:get(url, ADDITIONAL_HEADERS, retry_fn(3))
+  if (err ~= nil) then
+      client:shutdown()
+      self.client = RestClient.new("https://"..self.ip, HttpClient.labeled_socket_builder(self.wiser_index_code))
+  end
+  return process_rest_response(response,err,partial)
+end
+
+function HttpClient:do_post(url, content)
+  -- get url
+  local client = self.client
+  if (client == nil) then
```

Comment:
It seems like there should be an early return in the case client is nil and not just a warn log as you use `client` below.